Support splitting a single item into multiple files

missingdays opened the pull request:
## Motivation

To allow generating code for a single entity in multiple files. See a related issue in `volo` - https://github.com/cloudwego/volo/issues/454

## Solution

Introduce a new `split` mode that can be enabled using the builder

missingdays committed:
Support splitting a single item into multiple files

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cloudwego/pilota?pullRequest=270) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cloudwego/pilota?pullRequest=270) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cloudwego/pilota?pullRequest=270) it.</sub>

PureWhiteWu commented on pilota-build/src/lib.rs:
```diff
@@ -161,6 +165,11 @@ impl<MkB, P> Builder<MkB, P> {
         self
     }
 
+    pub fn with_split(mut self) -> Self {
```
Maybe we can add an bool arg like `change_case` below? In which case we can both opt-in or opt-out this feature.

PureWhiteWu reviewed (commented):
Thanks very much for this contribution!

I think we should add the type prefix to those generated files, since different types may not be in the same resolve namespace. For example, we can have a `service A` and `struct A` at the same time.

missingdays committed:
Rename `with_split` to `split_generated_files`, provide an bool parameter to it

missingdays committed:
Add item type to a generated file name

missingdays committed:
Linter fix

Millione commented:
LGTM. Maybe a test for the workspace mode can be added?

missingdays commented:
@Millione Hi. I tried adding a test for a workspace, but there are no other tests for it. And my test fails with a panic at `pilota-build/src/codegen/workspace.rs:91` because `Cargo.toml` doesn't exist. I'm not sure how this file should be created when generating the workspace for the first time. Here's how I compile with the workspace mode

```rust
      crate::Builder::thrift()
          .ignore_unused(false)
          .compile_with_config(
              vec![IdlService::from_path(source.to_owned())],
              crate::Output::Workspace(target.into()),
          )
```

Could you please point me to an existing test for a workspace mode, or help me figure out why `Cargo.toml` is expected to already exist in `group_defs`?

Millione commented:
> Could you please point me to an existing test for a workspace mode

Unfortunately, there is no existing test for this, but I'll do my best to help you figure it out.

> Why Cargo.toml is expected to already exist in group_defs?

The difference between normal mode and workspace mode is that in workspace mode the code generation doesn't take place in the build phase, but in the run phase. It generates multiple crates and adds all these crates as workspace members to the Cargo.toml file. 

> I'm not sure how this file should be created when generating the workspace for the first time

In my opinion, we should put the workspace mode test in a separate tests folder, so that it is a package with a binary target and fits into the real usage.

missingdays committed:
Add workspace tests

missingdays commented:
I added tests for the workspace mode, as well as splitting the files in the workspace mode.
I added all the tests from the thrift tests, I'm not sure if so many are needed or if it's too much

Millione commented:
> I added tests for the workspace mode, as well as splitting the files in the workspace mode.

Fantastic jobs!

> I added all the tests from the thrift tests, I'm not sure if so many are needed or if it's too much

Maybe that's a bit too much? I think one thrift test is enough, but the tested thrift files should be well-constructed for the workspace, e.g. A.thrift and B.thrift both use C.thrift struct, and see if the struct is generated in common crates and used correctly by A and B.

PureWhiteWu commented:
> A.thrift and B.thrift both use C.thrift struct

Maybe more complex? for example, more layers of nesting mods/namespaces?

missingdays commented:
Could you maybe write down the `.thrift` file you have in mind? My knowledge of thrift is quite limited

PureWhiteWu commented:
Sorry for the late reply, I've constructed five idl files.

article.thrift:

```thrift
include "image.thrift"
include "author.thrift"
include "common.thrift"

namespace rs article

enum Status {
    NORMAL = 0,
    DELETED = 1,
}

struct Article {
    1: required i64 id,
    2: required string title,
    3: required string content,
    4: required author.Author author,
    5: required Status status,
    6: required list<image.Image> images,
    7: required common.CommonData common_data,
}

struct GetArticleRequest {
    1: required i64 id,
}

struct GetArticleResponse {
    1: required Article article,
}

service ArticleService {
    GetArticleResponse GetArticle(1: GetArticleRequest req),
}
```

author.thrift:

```thrift
include "image.thrift"
include "common.thrift"

namespace rs author

struct Author {
    1: required i64 id,
    2: required string username,
    3: required string email,
    4: required image.Image avatar,
    5: required common.CommonData common_data,
}

struct GetAuthorRequest {
    1: required i64 id,
}

struct GetAuthorResponse {
    1: required Author author,
}

service AuthorService {
    GetAuthorResponse GetAuthor(1: GetAuthorRequest req),
}
```

cdn.thrift:

```thrift
include "common.thrift"

namespace rs article.image.cdn

struct CDN {
    1: required i64 id,
    2: required string url,
    3: required common.CommonData common_data,
}
```

common.thrift:

```thrift
namespace rs common

struct CommonData {
    1: required i64 id,
    2: required string name,
    3: required string description,
}
```

image.thrift:

```thrift
include "common.thrift"
include "cdn.thrift"

namespace rs article.image

struct Image {
    1: required i64 id,
    2: required string url,
    3: required cdn.CDN cdn,
    4: required common.CommonData common_data,
}

struct GetImageRequest {
    1: required i64 id,
}

struct GetImageResponse {
    1: required Image image,
}

service ImageService {
    GetImageResponse GetImage(1: GetImageRequest req),
}
```

And there are three services that can be used in workspace mode, in `article.thrift`, `author.thrift`, `image.thrift`.

Millione commented on pilota-build/test_data/thrift_workspace_with_split/output/image/src/lib.rs:
It seems not right?

Millione commented on pilota-build/test_data/thrift_workspace_with_split/output/image/src/message_ImageServiceGetImageArgsRecv.rs:
It seems that the `ImageService` trait not generated?

Millione commented on pilota-build/test_data/thrift_workspace_with_split/output/article/src/message_Author.rs:
If article.thrift has its own `Author` struct, will the filename conflict?

PureWhiteWu replied:
In the [original design](https://github.com/cloudwego/volo/issues/454#issuecomment-2338143899), I think this file should be placed under a new directory called `author`(same as the mod name) to avoid name conflict.

PureWhiteWu commented on pilota-build/test_data/thrift_workspace_with_split/output/article/src/message_Author.rs:
```diff
@@ -0,0 +1 @@
+pub use ::common::author::Author;
```
why do we need to `pub use` here instead of only `use`? @Millione

Millione replied:
Users can avoid importing the common crate dependencies

missingdays committed:
Fix tests

PureWhiteWu commented on pilota-build/test_data/thrift_workspace_with_split/output/article/src/gen.rs:
```diff
@@ -0,0 +1,32 @@
+pub mod gen {
+    #![allow(warnings, clippy::all)]
+
+    pub mod article {
```
I think every mod should be placed under a new directory as Rust does in the original design to avoid file name conflict and too many files under one directory.

missingdays committed:
Generate different mods in different directories

PureWhiteWu commented on pilota-build/test_data/thrift_workspace_with_split/output/article/src/gen.rs:
```diff
@@ -0,0 +1,32 @@
+    #![allow(warnings, clippy::all)]
+
+    pub mod article {
+        include!("article/enum_Status.rs");
+        include!("article/message_ArticleServiceGetArticleArgsRecv.rs");
+        include!("article/enum_ArticleServiceGetArticleResultSend.rs");
+        include!("article/message_GetArticleResponse.rs");
+        include!("article/message_ArticleServiceGetArticleArgsSend.rs");
+        include!("article/message_GetArticleRequest.rs");
+        include!("article/service_ArticleService.rs");
+        include!("article/enum_ArticleServiceGetArticleResultRecv.rs");
+        include!("article/message_Article.rs");
+
+        pub mod image {
+            include!("article/image/message_Image.rs");
+
+            pub mod cdn {
+                include!("article/image/cdn/message_CDN.rs");
+            }
+        }
```
Great! Thanks very much for your quick fix!

My last feature request: could we make a more thorough split like the original design?

For example, we can put these all in `article/mod.rs` and include it, which contains the following content:

```rust
include!("article/enum_Status.rs");
include!("article/message_ArticleServiceGetArticleArgsRecv.rs");
include!("article/enum_ArticleServiceGetArticleResultSend.rs");
include!("article/message_GetArticleResponse.rs");
include!("article/message_ArticleServiceGetArticleArgsSend.rs");
include!("article/message_GetArticleRequest.rs");
include!("article/service_ArticleService.rs");
include!("article/enum_ArticleServiceGetArticleResultRecv.rs");
include!("article/message_Article.rs");

pub mod image {
    include!("article/image/mod.rs"); // recursively
}
```

I'm not sure if this will add extra cost to the IDEs, but I think this is a more rusty way and is much more extensible.

What do you think? @missingdays @Millione

Millione replied:
Agreed

missingdays committed:
Add a check for successful `cargo build` when testing workspace mode

missingdays committed:
Generate `mod.rs` file for each mode when splitting

PureWhiteWu commented:
Thanks again! LGTM now, seems there's still a little format issue, could you please use `cargo fmt` to fix it?

missingdays committed:
Apply fmt fixes

PureWhiteWu commented:
LGTM, @Millione PTAL

missingdays commented:
Great, thanks for the reviews!
Now I'd like to use this functionality in `volo` to fix https://github.com/cloudwego/volo/issues/454
Could you please publish the new (unstable?) version of `pilota` so that I can use it in `volo`?

PureWhiteWu commented:
@Millione Maybe we can publish a new version and make split the default behaviour?

Millione commented:
> @Millione Maybe we can publish a new version and make split the default behaviour?

Get it.

Millione commented on pilota-build/test_data/thrift_workspace_with_split/output/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[workspace]
+members = [
+    "article",
+    "author", "common",
```
When I use the `UPDATE_TEST_DATE=1 cargo test` command, the common in members will disappear. Does it the same for you? @missingdays

Millione commented:
> Could you please publish the new (unstable?) version of pilota so that I can use it in volo?

@missingdays Maybe you can use the git dependencies temporarily? Once it is ready in volo, we can release the version in pilota.

missingdays commented:
@Millione @PureWhiteWu I have submite the pull request to `volo` with this functionality https://github.com/cloudwego/volo/pull/510
